File: algorithms/easy/count_pairs_of_similar_strings.cpp
class Solution {
public:
  int similarPairs(vector<string> &words) {
    unordered_map<int, int> mp;
    for (auto &w : words) {
      int res = 0;
      for (auto c : w) {
        res |= 1 << (c - 'a');
      }
      mp[res]++;
    }
    int ans = 0;
    for (auto &it : mp) {
      ans += it.second * (it.second - 1) / 2;
    }
    return ans;
  }
};

File: algorithms/easy/count_pairs_of_similar_strings.ex
defmodule Solution do
  use Bitwise

  @spec similar_pairs(words :: [String.t()]) :: integer
  def similar_pairs(words) do
    words
    |> Enum.reduce(%{}, fn w, acc ->
      w
      |> String.to_charlist()
      |> Enum.reduce(0, fn c, acc ->
        acc ||| 1 <<< (c - ?a)
      end)
      |> then(fn v -> Map.update(acc, v, 1, &(&1 + 1)) end)
    end)
    |> Enum.reduce(0, fn {_, v}, acc ->
      acc + div(v * (v - 1), 2)
    end)
  end
end
